From 941e6605241f9069bba4bc0f99e71c1fc234b08a Mon Sep 17 00:00:00 2001
From: Jayabalaji <jayabalaji.sathiyamoorthi@intel.com>
Date: Thu, 27 Feb 2025 14:52:43 +0100
Subject: [PATCH 04/14] Standalone debian packages for ORB-SLAM3 and examples
 (#5)

* Standalone debian packages for ORB-SLAM3 and examples

Signed-off-by: Sathiyamoorthi, Jayabalaji <jayabalaji.sathiyamoorthi@intel.com>

* Update control

* Update control

* Add pangolin library to fulfill the build condition

Signed-off-by: Sathiyamoorthi, Jayabalaji <jayabalaji.sathiyamoorthi@intel.com>

* Remove the Pangolin repo and build it on the fly

Signed-off-by: Sathiyamoorthi, Jayabalaji <jayabalaji.sathiyamoorthi@intel.com>

* Minor tweaks in the build script and dpkg build

Signed-off-by: Sathiyamoorthi, Jayabalaji <jayabalaji.sathiyamoorthi@intel.com>

* Added debian/orb-slam3.substvars contains the correct library paths.
Added dh_shlibdeps explicitly in the rules file to ensure the library paths are considered.
Added verbosity to dpkg-shlibdeps to see more output.
Added LDD in build.sh script to confirm proper installation of the Pangolin
Clean the build and rebuild it using dh_clean and dpkg-buildpackage.

Signed-off-by: Sathiyamoorthi, Jayabalaji <jayabalaji.sathiyamoorthi@intel.com>

---------

Signed-off-by: Sathiyamoorthi, Jayabalaji <jayabalaji.sathiyamoorthi@intel.com>
---
 .github/workflows/build.yml  |  3 +-
 build.sh                     | 19 ++++++--
 debian/changelog             |  2 +-
 debian/control               | 19 +++++---
 debian/orb-slam3-dev.install |  1 +
 debian/orb-slam3.install     |  1 +
 debian/orb-slam3.substvars   |  3 ++
 debian/rules                 | 86 ++++++++++--------------------------
 8 files changed, 61 insertions(+), 73 deletions(-)
 create mode 100644 debian/orb-slam3-dev.install
 create mode 100644 debian/orb-slam3.install
 create mode 100644 debian/orb-slam3.substvars

diff --git a/.github/workflows/build.yml b/.github/workflows/build.yml
index 834f46d..5009aa4 100644
--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -30,8 +30,9 @@ jobs:
 
       - name: Build
         run: |
-          dpkg-buildpackage
+          dpkg-buildpackage -us -uc
 
+      
       - name: Deploy Development
         shell: bash
         run: |
diff --git a/build.sh b/build.sh
index 96d1c09..1c0f00a 100755
--- a/build.sh
+++ b/build.sh
@@ -24,11 +24,24 @@ cd build
 cmake .. -DCMAKE_BUILD_TYPE=Release
 make -j
 
-cd ../../../
+cd ../../../Thirdparty
+echo "Cloning Pangolin repo..."
+git clone --recursive https://github.com/stevenlovegrove/Pangolin.git
+cd Pangolin
+echo "Installing prerequisites..."
+dependencies=$(./scripts/install_prerequisites.sh --list)
+apt-get install -y ${dependencies}
+echo "Configuring and building Thirdparty/Pangolin ..."
+mkdir build
+cd build
+cmake .. -DCMAKE_BUILD_TYPE=Release
+make -j
+make install 
+echo "ldd pangolin.."
+ldd /usr/local/lib/libpango_display.so.0
 
+cd ../../../Vocabulary
 echo "Uncompress vocabulary ..."
-
-cd Vocabulary
 tar -xf ORBvoc.txt.tar.gz
 cd ..
 
diff --git a/debian/changelog b/debian/changelog
index 79e2790..66899b3 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,4 +1,4 @@
-ros-humble-orb-slam3 (1.0-1) UNRELEASED; urgency=low
+orb-slam3 (1.0-1) UNRELEASED; urgency=low
 
   * Initial release
 
diff --git a/debian/control b/debian/control
index 57027ce..8b8fc26 100644
--- a/debian/control
+++ b/debian/control
@@ -1,12 +1,21 @@
-Source: ros-humble-orb-slam3
+Source: orb-slam3
 Section: utils
 Priority: optional
 Maintainer: ECI Maintainer <eci.maintainer@intel.com>
-Build-Depends: cmake, debhelper (>=11~), ros-humble-sophus, ros-humble-pangolin, ros-humble-libg2o-intel-avx2, ros-humble-ament-cmake, ros-humble-ament-lint-common, ros-humble-ament-lint-auto, ros-humble-ament-cmake-gtest, ros-humble-std-msgs, ros-humble-geometry-msgs, ros-humble-sensor-msgs, ros-humble-visualization-msgs, ros-humble-nav-msgs, ros-humble-tf2-geometry-msgs, ros-humble-statistics-msgs, ros-humble-cv-bridge, ros-humble-rosbag2-cpp, ros-humble-rosbag2-transport, ros-humble-tf2-ros, ros-humble-message-filters, ros-humble-cv-bridge, ros-humble-image-transport, ros-humble-image-transport-plugins, ros-humble-dbow2, libspdlog-dev, nlohmann-json3-dev, libboost-dev, python3-ament-package, libopencv-dev, libeigen3-dev, ros-humble-rclcpp, libgtest-dev, ros-humble-ament-cmake-gtest
+Build-Depends: debhelper (>= 12), cmake (>= 3.10), g++ (>= 8), libopencv-dev, libeigen3-dev, libboost-all-dev, libsdl2-dev, libssl-dev, libprotobuf-dev, librealsense2-dev, python3-wheel
 Standards-Version: 4.1.4
 
-Package: ros-humble-orb-slam3
+Package: orb-slam3
 Architecture: any
 Multi-Arch: foreign
-Depends: ${misc:Depends}, ${shlibs:Depends}, ros-humble-nav2-dynamic-msgs, ros-humble-builtin-interfaces, ros-humble-geometry-msgs, ros-humble-visualization-msgs, ros-humble-sensor-msgs, ros-humble-rclcpp, ros-humble-rviz2, ros-humble-robotics-sdk-license
-Description: ORB-SLAM3: An Accurate Open-Source Library for Visual, Visual-Inertial and Multi-Map SLAM 
+Depends: ${misc:Depends}, ${shlibs:Depends}, libopencv4.4, libeigen3.1.0, libboost-all-dev, libssl-dev, librealsense2-dev
+Description: ORB-SLAM3: A robust and efficient SLAM system
+ ORB-SLAM3 is a state-of-the-art Visual SLAM system for monocular, stereo, and RGB-D cameras, along with inertial 
+ measurement units (IMU). It builds and maintains a 3D map of the environment using various sensor inputs and performs 
+ real-time localization and mapping.
+
+Package: orb-slam3-dev
+Section: libdevel
+Architecture: any
+Depends: orb-slam3 (= ${binary:Version}), ${misc:Depends}
+Description: ORB SLAM3 development headers
diff --git a/debian/orb-slam3-dev.install b/debian/orb-slam3-dev.install
new file mode 100644
index 0000000..8787c1b
--- /dev/null
+++ b/debian/orb-slam3-dev.install
@@ -0,0 +1 @@
+usr/include
\ No newline at end of file
diff --git a/debian/orb-slam3.install b/debian/orb-slam3.install
new file mode 100644
index 0000000..0d5487b
--- /dev/null
+++ b/debian/orb-slam3.install
@@ -0,0 +1 @@
+usr/lib
\ No newline at end of file
diff --git a/debian/orb-slam3.substvars b/debian/orb-slam3.substvars
new file mode 100644
index 0000000..8d8fb32
--- /dev/null
+++ b/debian/orb-slam3.substvars
@@ -0,0 +1,3 @@
+libpango_display.so.0 /usr/local/lib/libpango_display.so.0
+libDBoW2.so /usr/local/lib/libDBoW2.so
+libg2o.so /usr/local/lib/libg2o.so
\ No newline at end of file
diff --git a/debian/rules b/debian/rules
index a14b720..55d7210 100755
--- a/debian/rules
+++ b/debian/rules
@@ -1,72 +1,32 @@
 #!/usr/bin/make -f
-# -*- makefile -*-
-# Sample debian/rules that uses debhelper.
-# This file was originally written by Joey Hess and Craig Small.
-# As a special exception, when this file is copied by dh-make into a
-# dh-make output file, you may use that output file without restriction.
-# This special exception was added by Craig Small in version 0.37 of dh-make.
 
-# Uncomment this to turn on verbose mode.
-export DH_VERBOSE=1
-# TODO: remove the LDFLAGS override.  It's here to avoid esoteric problems
-# of this sort:
-#  https://code.ros.org/trac/ros/ticket/2977
-#  https://code.ros.org/trac/ros/ticket/3842
-export LDFLAGS=-pthread
-export PKG_CONFIG_PATH=/opt/ros/humble/lib/pkgconfig
-# Explicitly enable -DNDEBUG, see:
-# 	https://github.com/ros-infrastructure/bloom/issues/327
-export DEB_CXXFLAGS_MAINT_APPEND=-DNDEBUG
-ifneq ($(filter nocheck,$(DEB_BUILD_OPTIONS)),)
-       BUILD_TESTING_ARG=-DBUILD_TESTING=OFF
-endif
-export DEB_CXXFLAGS_MAINT_APPEND=-Wno-error=maybe-uninitialized
-# export DEB_LDFLAGS_MAINT_APPEND = -lpthread
+export DH_VERBOSE = 1
+export DEB_BUILD_MAINT_OPTIONS = hardening=+all
+export DEB_HOST_MULTIARCH := $(shell dpkg-architecture -qDEB_HOST_MULTIARCH)
+export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
 
-DEB_HOST_GNU_TYPE ?= $(shell dpkg-architecture -qDEB_HOST_GNU_TYPE)
-
-export ROS_DISTRO=humble
 
 %:
-	dh $@ -v --buildsystem=cmake --parallel --builddirectory=build \
-	--sourcedirectory=./
+	dh $@ --parallel
 
+# Override the default configure step to include custom third-party library builds
 override_dh_auto_configure:
-	# In case we're installing to a non-standard location, look for a setup.sh
-	# in the install tree and source it.  It will set things like
-	# CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
-	if [ -f "/opt/ros/humble/setup.sh" ]; then . "/opt/ros/humble/setup.sh"; fi && \
-	dh_auto_configure -- \
-		-DCMAKE_INSTALL_PREFIX=/opt/ros/humble \
-		-DAMENT_PREFIX_PATH=/opt/ros/humble \
-		-DCMAKE_PREFIX_PATH=/opt/ros/humble \
-		 -DCMAKE_BUILD_TYPE=Release
-
-override_dh_auto_build:
-	# In case we're installing to a non-standard location, look for a setup.sh
-	# in the install tree and source it.  It will set things like
-	# CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
-	if [ -f "/opt/ros/humble/setup.sh" ]; then . "/opt/ros/humble/setup.sh"; fi && \
-	dh_auto_build
-
-override_dh_auto_test:
-	# In case we're installing to a non-standard location, look for a setup.sh
-	# in the install tree and source it.  It will set things like
-	# CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
-	echo -- Running tests. Even if one of them fails the build is not canceled.
-	if [ -f "/opt/ros/humble/setup.sh" ]; then . "/opt/ros/humble/setup.sh"; fi && \
-	dh_auto_test || true
+# Building Thirdparty dependencies (DBoW2, g2o, Sophus, Pangolin) and run build script
+	./build.sh
 
 override_dh_shlibdeps:
-	# In case we're installing to a non-standard location, look for a setup.sh
-	# in the install tree and source it.  It will set things like
-	# CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
-	if [ -f "/opt/ros/humble/setup.sh" ]; then . "/opt/ros/humble/setup.sh"; fi && \
-	dh_shlibdeps -l$(CURDIR)/debian/ros-humble-orb-slam3//opt/ros/humble/lib/
-
-override_dh_auto_install:
-	# In case we're installing to a non-standard location, look for a setup.sh
-	# in the install tree and source it.  It will set things like
-	# CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
-	if [ -f "/opt/ros/humble/setup.sh" ]; then . "/opt/ros/humble/setup.sh"; fi && \
-	dh_auto_install
+	dh_shlibdeps --dpkg-shlibdeps-params=--ignore-missing-info
+
+# Override the default install step to handle custom installation paths and files
+override_dh_install:
+# Install libraries and headers
+	install -d debian/orb-slam3/usr/lib/$(DEB_HOST_MULTIARCH)
+	install -d debian/orb-slam3/usr/include/orb-slam3
+	ls -lah
+	install -m 0755 lib/libORB_SLAM3.so debian/orb-slam3/usr/lib/$(DEB_HOST_MULTIARCH)
+	cp -r include/* debian/orb-slam3/usr/include/orb-slam3
+
+# Clean up any temporary files that were generated during the build
+override_dh_clean:
+# Perform cleanup after the build
+	dh_clean
\ No newline at end of file
-- 
2.30.2

